File: ui/web/src/types/state.ts
export type Vector3 = {
  x: number;
  y: number;
  z: number;
};

export type Quaternion = {
  x: number;
  y: number;
  z: number;
  w: number;
};

export type EstimatedState = {
  timestamp_us: number;
  position_m: Vector3;
  velocity_mps: Vector3;
  orientation_xyzw: Quaternion;
  uncertainty: {
    sigma_x_m: number;
    sigma_y_m: number;
    sigma_z_m: number;
  };
  status: {
    zupt_active: boolean;
    uwb_available: boolean;
    active_anchor_count: number;
  };
};

export type StateMessage = {
  type: "estimated_state";
  source: "simulation" | "live" | "replay";
  state: EstimatedState;
};

export type Anchor = {
  anchor_id: string;
  x_m: number;
  y_m: number;
  z_m: number;
};

File: backend/models/state.py
from __future__ import annotations

from dataclasses import asdict, dataclass


@dataclass(frozen=True)
class Vector3:
    x: float
    y: float
    z: float


@dataclass(frozen=True)
class Quaternion:
    x: float
    y: float
    z: float
    w: float


@dataclass(frozen=True)
class Uncertainty:
    sigma_x_m: float
    sigma_y_m: float
    sigma_z_m: float


@dataclass(frozen=True)
class SensorStatus:
    zupt_active: bool
    uwb_available: bool
    active_anchor_count: int


@dataclass(frozen=True)
class EstimatedState:
    timestamp_us: int
    position_m: Vector3
    velocity_mps: Vector3
    orientation_xyzw: Quaternion
    uncertainty: Uncertainty
    status: SensorStatus

    def to_dict(self) -> dict:
        return asdict(self)


def websocket_message(state: EstimatedState, source: str = "simulation") -> dict:
    return {
        "type": "estimated_state",
        "source": source,
        "state": state.to_dict(),
    }

